cngi/dio/read_image.py

```diff
@@ -34,20 +36,68 @@ def read_image(infile, chunks=None, consolidated=True, overwrite_encoded_chunks=
     overwrite_encoded_chunks : bool
         drop the zarr chunks encoded for each variable when a dataset is loaded with specified chunk sizes.  Default True, only applies when chunks
         is not None.
-  
+    s3_key : string, optional
+        optional support for explicit authentication if infile is provided as S3 URL. 
+        If S3 url is passed as input but this argument is not specified then only publicly-available, read-only buckets are accessible (so output dataset will be read-only).
+    s3_secret : string, optional
+        optional support for explicit authentication if infile is provided as S3 URL. 
+        If S3 url is passed as input but this argument is not specified then only publicly-available, read-only buckets are accessible (so output dataset will be read-only).  
+
     Returns
     -------
     xarray.core.dataset.Dataset
         New xarray Dataset of image contents
     """
     import os
     from xarray import open_zarr
- 
+
     if chunks is None:
-        chunks = 'auto'
+        chunks = "auto"
         overwrite_encoded_chunks = False
- 
-    infile = os.path.expanduser(infile)
-    xds = open_zarr(infile, chunks=chunks, consolidated=consolidated, overwrite_encoded_chunks=overwrite_encoded_chunks)
-    return xds
 
+    if infile.lower().startswith("s3"):
```

Comment:
What about splitting the "s3" case into a separate function? This is to keep individual functions small and focused on one thing, and also a more explicit separation of what parameters are needed in which case.
And taking one more step, if we also put the "non-s3" case into its own function then it becomes trivially clear that this top level function handles the different types of read. And then we can give an easy to read name to each of those functions. As simple as it sounds this is incredibly helpful in making code "clean" and simple/quick/trivial to read.
This is not very relevant at this stage of prototyping, but longer term It would be good to agree on some general approx. limits on function length, number of parameters, complexity, etc.

Reply:
We went back and forth when considering whether it was right to combine this under one function or have the use cases decoupled at the interface level. I agree that your idea seems like the right approach from a readability and maintainability standpoint, and support separating into multiple functions called by a parent upon further development.